XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs

```diff
@@ -125,205 +124,279 @@ protected override void OnDropDownOpening(EventArgs e)
             {
                 SelectedItemCollection selection = Command.GetSelection();
                 Session session = selection[0].Connection.DuplicateSession();
-                if (Helpers.WlbEnabled(selection[0].Connection))
-                {
-                    WlbRecommendations recommendations = new WlbRecommendations(selection.AsXenObjects<VM>(), session);
-                    recommendations.Initialize();
 
-                    if (recommendations.IsError)
-                        EnableAppropriateHostsNoWlb(session);
-                    else
-                        EnableAppropriateHostsWlb(session, recommendations);
-                }
-                else
-                {
-                    EnableAppropriateHostsNoWlb(session);
-                }
+                hostListUpdater.updateHostList(this, session);
+                
             });
         }
 
-        private void EnableAppropriateHostsWlb(Session session, WlbRecommendations recommendations)
+        /// <summary>
+        /// Hook to add additional members to the menu item
+        /// Note: Called on main window thread by executing code
+        /// </summary>
+        /// <param name="selection"></param>
+        protected virtual void AddAdditionalMenuItems(SelectedItemCollection selection) { return; }
+
+        private class HostListUpdater
         {
-            SelectedItemCollection selection = Command.GetSelection();
-            // set the first menu item to be the WLB optimal server menu item
-            VMOperationToolStripMenuSubItem firstItem = (VMOperationToolStripMenuSubItem)base.DropDownItems[0];
-            var firstItemCmd = new VMOperationWlbOptimalServerCommand(Command.MainWindowCommandInterface, selection, _operation, recommendations);
-            var firstItemCmdCanExecute = firstItemCmd.CanExecute();
+            private ProduceConsumerQueue workQueue = new ProduceConsumerQueue(1);
+            private ProduceConsumerQueue workerQueueWithouWlb = new ProduceConsumerQueue(25);

```

Comment:
I believe this should have been called workerQueueWithoutWlb (missing 't')

Reply:
workerQueueWithouWlb is always recreated in updateHostList, so I don't think it is necessary to be created here as well, especially as the ProduceConsumerQueue constructor creates and starts 25 threads.